File: minerId-generator.js
// Distributed under the Open BSV software license, see the accompanying file LICENSE.

const bsv = require('bsv')

const protocolName = '601dface'
const cbdVersion = '0.3'
const protocolIdVersion = '00'

function createMinerInfoDocument (height, minerIdPublicKey, prevMinerIdPrivKey, revocationKeyPublicKey, prevRevocationKeyPrivKey, optionalData) {
  let prevMinerIdPublicKey = prevMinerIdPrivKey.toPublicKey().toString()

  prevMinerIdPublicKey = prevMinerIdPublicKey || minerIdPublicKey
  const minerIdSigPayload = Buffer.concat([
    Buffer.from(prevMinerIdPublicKey, 'hex'),
    Buffer.from(minerIdPublicKey, 'hex')
  ])
  const hash = bsv.crypto.Hash.sha256(minerIdSigPayload)
  const prevMinerIdSig = bsv.crypto.ECDSA.sign(hash, prevMinerIdPrivKey).toString()

  let prevRevocationKeyPublicKey = prevRevocationKeyPrivKey.toPublicKey().toString()
  prevRevocationKeyPublicKey = prevRevocationKeyPublicKey || revocationKeyPublicKey
  const prevRevocationKeySigPayload = Buffer.concat([
    Buffer.from(prevRevocationKeyPublicKey, 'hex'),
    Buffer.from(revocationKeyPublicKey, 'hex')
  ])
  const hash2 = bsv.crypto.Hash.sha256(prevRevocationKeySigPayload)
  const prevRevocationKeySig = bsv.crypto.ECDSA.sign(hash2, prevRevocationKeyPrivKey).toString()

  const doc = {
    version: cbdVersion,
    height: height,

    prevMinerId: prevMinerIdPublicKey,
    prevMinerIdSig: prevMinerIdSig,

    minerId: minerIdPublicKey,

    prevRevocationKey: prevRevocationKeyPublicKey,
    prevRevocationKeySig: prevRevocationKeySig,

    revocationKey: revocationKeyPublicKey,
  }
  if (optionalData) {
    doc.minerContact = optionalData
  }
  return doc
}

function createMinerInfoOpReturn (height, minerIdPrivKey, prevMinerIdPrivKey, revocationKeyPrivKey, prevRevocationKeyPrivKey, mc) {
  const minerIdPublicKey = minerIdPrivKey.toPublicKey().toString()
  const revocationKeyPublicKey = revocationKeyPrivKey.toPublicKey().toString()
  const doc = createMinerInfoDocument(height, minerIdPublicKey, prevMinerIdPrivKey, revocationKeyPublicKey, prevRevocationKeyPrivKey, mc)

  const payload = JSON.stringify(doc)

  const hash = bsv.crypto.Hash.sha256(Buffer.from(payload))
  const signature = bsv.crypto.ECDSA.sign(hash, minerIdPrivKey).toString()

  const opReturnScript = bsv.Script.buildSafeDataOut([protocolName, protocolIdVersion, payload, signature]).toHex()
  return opReturnScript
}

function createMinerInfoCoinbaseTxOpReturn (minerInfoTxId, blockBind, blockBindSig) {
  const minerInfoTxIdInLittleEndianRep = Buffer.from(minerInfoTxId, 'hex').reverse() // swap endianness before adding into the script
  return bsv.Script.buildSafeDataOut([protocolName, protocolIdVersion, minerInfoTxIdInLittleEndianRep, blockBind, blockBindSig]).toHex()
}

const height = 123
const mc = {
  name: 'demo',
  email: 'demo@demo.com',
  merchantAPIEndPoint: 'api.demo.com'
}
const prevMinerIdPrivKey = new bsv.PrivateKey()
const minerIdPrivKey = new bsv.PrivateKey()

const prevRevocationKeyPrivKey = new bsv.PrivateKey()
const revocationKeyPrivKey = new bsv.PrivateKey()

const minerInfoOpReturn = createMinerInfoOpReturn(height, minerIdPrivKey, prevMinerIdPrivKey, revocationKeyPrivKey, prevRevocationKeyPrivKey, mc)
console.log(`Miner-info op_return Output Script: ${minerInfoOpReturn}`)

const minerInfoTxId = "f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16"
const blockBind = "0028944a3a436201521bafa5cf82f873c04d212d62d5811324a1fd14095a6ea2"
const blockBindSig = "304402206ea641c5a1568d06572629ab46deef74b351d65e5d3112c9c24cecd896a1108c0220337ba129162c26e6aa996d1f88164566c03ee395d75a63033cb421fc432f1e7a"
const midCoinbaseOpReturn = createMinerInfoCoinbaseTxOpReturn(minerInfoTxId, blockBind, blockBindSig)
console.log(`Miner ID Coinbase Tx op_return Output Script: ${midCoinbaseOpReturn}`)

File: minerId-generator.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"

	"github.com/bitcoinsv/bsvd/bsvec"
	"github.com/libsv/libsv/transaction/output"
)

const protocolName = "ac1eed88"
const cbdVersion = "0.1"

type coinbaseDoc struct {
	Version        string       `json:"version"`
	Height         uint64       `json:"height"`
	PrevMinerID    string       `json:"prevMinerId"`
	PrevMinerIDSig string       `json:"prevMinerIdSig"`
	MinerID        string       `json:"minerId"`
	Vctx           vctx         `json:"vctx"`
	MinerContact   minerContact `json:"minerContact"`
}

type vctx struct {
	TxID string `json:"txId"`
	Vout uint64 `json:"vout"`
}

type minerContact struct {
	Name                string `json:"name"`
	Email               string `json:"email"`
	MerchantAPIEndPoint string `json:"merchantAPIEndPoint"`
}

func createCoinbaseDocument(height uint64, minerID string, prevMinerIDPrivKey *bsvec.PrivateKey, vcTx string, mc minerContact) (coinbaseDoc, error) {
	prevMinerID := prevMinerIDPrivKey.PubKey().SerializeCompressed()

	p1 := prevMinerID
	p2, _ := hex.DecodeString(minerID)
	p3, _ := hex.DecodeString(vcTx)

	payload := []byte{}
	payload = append(payload, p1...)
	payload = append(payload, p2...)
	payload = append(payload, p3...)

	hash := sha256.Sum256(payload)

	prevMinerIDSig, err := prevMinerIDPrivKey.Sign(hash[:])
	if err != nil {
		fmt.Println(err)
		return coinbaseDoc{}, err
	}

	v := vctx{
		TxID: vcTx,
		Vout: 0,
	}

	doc := coinbaseDoc{
		Version: cbdVersion,
		Height:  height,

		PrevMinerID:    hex.EncodeToString(prevMinerID),
		PrevMinerIDSig: hex.EncodeToString(prevMinerIDSig.Serialize()),

		MinerID:      minerID,
		Vctx:         v,
		MinerContact: mc,
	}

	return doc, nil
}

func createMinerIDOpReturn(height uint64, prevMinerIDPrivKey *bsvec.PrivateKey, minerIDPrivKey *bsvec.PrivateKey, vcTx string, mc minerContact) (string, error) {
	minerID := hex.EncodeToString(minerIDPrivKey.PubKey().SerializeCompressed())

	doc, err := createCoinbaseDocument(height, minerID, prevMinerIDPrivKey, vcTx, mc)
	if err != nil {
		return "", err
	}

	payload, err := json.Marshal(doc)
	if err != nil {
		return "", err
	}
	hash := sha256.Sum256([]byte(payload))
	signature, err := minerIDPrivKey.Sign(hash[:])
	if err != nil {
		return "", err
	}

	var parts [][]byte
	// 00
	parts = append(parts, []byte(protocolName))
	parts = append(parts, payload)
	parts = append(parts, []byte(hex.EncodeToString(signature.Serialize())))

	// If we have some data, make another output with an OP_RETURN for it.
	output, err := output.NewOpReturnParts(parts)
	// output, err := transaction.NewOutputOpReturnPush(parts)
	if err != nil {
		return "", err
	}

	return output.GetLockingScriptHexString(), nil
}

func main() {
	var h uint64 = 123
	// mid := "03f9d057f6ff6606f615302812db37f6eccdd531cdd2c321673af187cf7dbbb9ae"
	v := "11c9f0be55da88192f1b6538468975bcfc1635c48f1ce9eeae12cdaefc5a4c99"
	privKey, err := bsvec.NewPrivateKey(bsvec.S256())
	if err != nil {
		fmt.Println(err)
		return
	}
	prevPrivKey, err := bsvec.NewPrivateKey(bsvec.S256())
	if err != nil {
		fmt.Println(err)
		return
	}
	mc := minerContact{
		Name:                "demo",
		Email:               "demo@demo.com",
		MerchantAPIEndPoint: "api.demo.com",
	}

	// doc, err := createCoinbaseDocument(h, mid, privKey, v, mc)

	opr, err := createMinerIDOpReturn(h, prevPrivKey, privKey, v, mc)

	// fmt.Println(doc)
	fmt.Printf("%+v\n", opr)
}
